### ISSUE TITLE
Math.exp(10_000.0) returns 0.0 instead of Infinity with JDK 14

### ISSUE THREAD

**Issue by schmittjoaopedro** [AdoptOpenJDK/openjdk-support#182]
## Summary

I'm executing some calculations using Java x86 in a Windows 64 (due to an external DLL library that requires x86 Java for windows environments) and I've got stuck with the following problem.

```java
System.out.println(Math.exp(10_000.0)); // Prints 0.0
```

I'm using the AdoptJDK version "14.0.2" (x86) (build 14.0.2+12) downloaded [here](https://adoptopenjdk.net/archive.html). I noticed the same error occurring since version openjdk 11.0.8 (x86) (build 11.0.8+10)
When using different versions like: 
> oracle 1.8.0_261 (x86) (build 1.8.0_261-b12)
> openjdk version "13" 2019-09-17 (x64)
> openjdk 14.0.2 2020-07-14 (x64)

The command below prints Infinity (that should be the right answer):

```java
System.out.println(Math.exp(10_000.0));
        // openjdk 14.0.2 (x86) (build 14.0.2+12) 2020-07-14 -> prints 0.0
        // openjdk 11.0.8 (x86) (build 11.0.8+10) 2020-07-14 -> prints 0.0
        // openjdk 14.0.2 (x64) (build 14.0.2+12) 2020-07-14 -> prints Infinity
        // oracle 1.8.0_261 (x86) (build 1.8.0_261-b12) -> prints Infinity
        // openjdk version "13" 2019-09-17 (x64) -> prints Infinity
```

## Triaging info

Java version (unziped from the downloaded package): 
```bash
C:\java\jdk-14.0.2+12-x86\bin>java.exe -version
openjdk version "14.0.2" 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 14.0.2+12)
OpenJDK Client VM AdoptOpenJDK (build 14.0.2+12, mixed mode, sharing)

C:\java\jdk-11.0.8+10-x86\bin>java.exe -version
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.8+10)
OpenJDK Client VM AdoptOpenJDK (build 11.0.8+10, mixed mode)
```

SO:
* Type: Windows 10 Home (x64)
* Version: 2004
* Compilation: 19041.508

**Comment by aahlenst** [comment]
Interesting.

Testprogram:
```
public class Test {
	public static void main(String[] args) {
		System.out.println(Math.exp(10_000.0));		
	}
}
```

On Linux:
```
$ javac Test.java 
$ java Test
Infinity
$ java -version
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.8+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.8+10, mixed mode)
```

Red Hat build on Windows:
```
PS C:\Users\Administrator\Downloads> .\openjdk-11.0.8_10\bin\javac.exe Test.java
PS C:\Users\Administrator\Downloads> .\openjdk-11.0.8_10\bin\java.exe Test
Infinity
PS C:\Users\Administrator\Downloads> .\openjdk-11.0.8_10\bin\java.exe -version
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment 18.9 (build 11.0.8+10)
OpenJDK 64-Bit Server VM 18.9 (build 11.0.8+10, mixed mode)
```

AdoptOpenJDK on Windows:
```
PS C:\Users\Administrator\Downloads> .\jdk-11.0.8+10\bin\javac.exe Test.java
PS C:\Users\Administrator\Downloads> .\jdk-11.0.8+10\bin\java.exe Test
Infinity
PS C:\Users\Administrator\Downloads> .\jdk-11.0.8+10\bin\java.exe -version
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.8+10)
OpenJDK 64-Bit Server VM AdoptOpenJDK (build 11.0.8+10, mixed mode)
```

AdoptOpenJDK x86 (32 bit) on Windows:
```
PS C:\Users\Administrator\Downloads> .\jdk-11.0.8+10\bin\javac.exe Test.java
PS C:\Users\Administrator\Downloads> .\jdk-11.0.8+10\bin\java.exe Test
0.0
PS C:\Users\Administrator\Downloads> .\jdk-11.0.8+10\bin\java.exe -version
openjdk version "11.0.8" 2020-07-14
OpenJDK Runtime Environment AdoptOpenJDK (build 11.0.8+10)
OpenJDK Client VM AdoptOpenJDK (build 11.0.8+10, mixed mode)
```

Happens with Azul Zulu, too:
```
PS C:\Users\Administrator\Downloads> .\zulu11.41.23-ca-jdk11.0.8-win_i686\bin\javac.exe Test.java
PS C:\Users\Administrator\Downloads> .\zulu11.41.23-ca-jdk11.0.8-win_i686\bin\java.exe Test
0.0
PS C:\Users\Administrator\Downloads> .\zulu11.41.23-ca-jdk11.0.8-win_i686\bin\java.exe -version
openjdk version "11.0.8" 2020-07-14 LTS
OpenJDK Runtime Environment Zulu11.41+23-CA (build 11.0.8+10-LTS)
OpenJDK Client VM Zulu11.41+23-CA (build 11.0.8+10-LTS, mixed mode)
```

Next question: What is a double in 32 bit Java and should it make a difference.

**Comment by schmittjoaopedro** [comment]
Sorry, I don't know if this question was for me, but let me give you more context about my scenario.

I'm working with interval arithmetics and I need to compute bounds for numerical values, I mean, in this kind of computation is very common to get stuck in bounds with very high values. For practical reasons I just the need the `Math.exp` bounding for Infinity. I don't think that a small word (32bits) will be a problem for my calculations. Additionally, I'm using JDK 1.8 32bits where Math.pow works fine, and my computation results are returned as expected.

As a developer I'm using the Windows SO (coorporative reasons), and because this solver requires a C++ library to compute interval arithmetic for real numbers, I need to use the JDK 32bits because I use MinGW plataform to generate the DLLs (dependent libraries).

**Comment by aahlenst** [comment]
@schmittjoaopedro Thanks for the additional info, but that was actually a note for me to remember where I left off.

Is using https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Double.html#POSITIVE_INFINITY a viable workaround?

**Comment by schmittjoaopedro** [comment]
Well, I think so.
Using JDK 8 (x86), the following listing prints true, then I think it could be a good workaround for OpenJDK.
```java
    public static void main(String[] args) {
        System.out.println(Math.exp(10_000.0) == Double.POSITIVE_INFINITY); // true
    }
```

**Comment by aahlenst** [comment]
```
PS C:\Users\Administrator\Downloads> .\jdk-15+36\bin\javac.exe Test.java
PS C:\Users\Administrator\Downloads> .\jdk-15+36\bin\java.exe Test
0.0
PS C:\Users\Administrator\Downloads> .\jdk-15+36\bin\java.exe -version
openjdk version "15" 2020-09-15
OpenJDK Runtime Environment AdoptOpenJDK (build 15+36)
OpenJDK Client VM AdoptOpenJDK (build 15+36, mixed mode)
```

```
PS C:\Users\Administrator\Downloads> .\jdk8u265-b01\bin\javac.exe Test.java
PS C:\Users\Administrator\Downloads> .\jdk8u265-b01\bin\java.exe Test
Infinity
PS C:\Users\Administrator\Downloads> .\jdk8u265-b01\bin\java.exe -version
openjdk version "1.8.0_265"
OpenJDK Runtime Environment (AdoptOpenJDK)(build 1.8.0_265-b01)
OpenJDK Client VM (AdoptOpenJDK)(build 25.265-b01, mixed mode)
```

Azul Zulu on Linux (32 bit mode):

```
$ zulu15.27.17-ca-jdk15.0.0-linux_i686/bin/javac Test.java 
$ zulu15.27.17-ca-jdk15.0.0-linux_i686/bin/java Test      
0.0
$ zulu15.27.17-ca-jdk15.0.0-linux_i686/bin/java --version
openjdk 15 2020-09-15
OpenJDK Runtime Environment Zulu15.27+17-CA (build 15+36)
OpenJDK Server VM Zulu15.27+17-CA (build 15+36, mixed mode)
```

Summary of my findings so far: JDK 8 (32 bit) considers `System.out.println(Math.exp(10_000.0));` to be Infinity. Somewhere between 8 and 11, that changed to 0.0, regardless of the operating system and language settings.

**Comment by aahlenst** [comment]
[Math#exp(double) delegates](https://github.com/openjdk/jdk/blob/83b2411fd15d0890042ede118a08731ca162b951/src/java.base/share/classes/java/lang/Math.java#L289) to StrictMath#exp(double) that [delegates to FdLibm.exp.compute(double)](https://github.com/openjdk/jdk/blob/83b2411fd15d0890042ede118a08731ca162b951/src/java.base/share/classes/java/lang/StrictMath.java#L230) that [is implemented in Java](https://github.com/openjdk/jdk/blob/83b2411fd15d0890042ede118a08731ca162b951/src/java.base/share/classes/java/lang/FdLibm.java#L684).

What could explain the difference is that with OpenJDK 9, `Math.exp()` was ported to Java: https://bugs.openjdk.java.net/browse/JDK-8134780.

**Comment by jerboaa** [comment]
What happens when run with `-XX:-UseLibmIntrinsic`?

**Comment by jerboaa** [comment]
So the bug is in the intrinsics code. Likely here:
http://hg.openjdk.java.net/jdk-updates/jdk11u-dev/file/360916af5610/src/hotspot/cpu/x86/stubGenerator_x86_32.cpp#l3375

**Comment by aahlenst** [comment]
https://mail.openjdk.java.net/pipermail/core-libs-dev/2020-September/069025.html

**Comment by aahlenst** [comment]
http://hg.openjdk.java.net/jdk-updates/jdk11u-dev/file/360916af5610/src/hotspot/cpu/x86/stubGenerator_x86_32.cpp#l3375 is only the stub. The actual assembler code for x86 is most likely https://github.com/openjdk/jdk/blob/7817963ce91384380fe65005a6c438b08b022bad/src/hotspot/cpu/x86/macroAssembler_x86_exp.cpp#L482, all generated by the Intel C compiler.

**Comment by aahlenst** [comment]
Fixed in 16: https://bugs.openjdk.java.net/browse/JDK-8255368. We'll backport to 11.

**Comment by aahlenst** [comment]
Backport has just landed in 11u-dev, should become part of 11.0.11 due in April.
